userspace/libsinsp/sinsp_filtercheck_event.cpp

```diff
@@ -561,13 +557,9 @@ uint8_t *sinsp_filter_check_event::extract_abspath(sinsp_evt *evt, OUT uint32_t
 		return 0;
 	}
 
-	parinfo = evt->get_param(dirfdargidx);
-	ASSERT(parinfo->m_len == sizeof(int64_t));
-	int64_t dirfd = *(int64_t *)parinfo->m_val;
+	int64_t dirfd = evt->get_param<int64_t>(dirfdargidx);
 
-	parinfo = evt->get_param(pathargidx);
-	path = parinfo->m_val;
-	pathlen = parinfo->m_len;
+	const char *path = evt->get_param_const_char(pathargidx);
```

Comment:
Here we get a small perf penalty because we then need to compute `strlen` ourselves, instead of using the one stored in `parinfo->m_len` right?

Reply:
Perhaps in this case `get_param` was not wrong and we can revert to previous code?

Reply:
Nothing gets past you :smile: However, this is deliberate.

I believe the correct signature for a function that extracts a C-style string is to return a `const char*`, optionally NULL if that is allowed. The memory that the string points to must also be safe to read until the terminator. Since C-style strings do not carry the length the length should not be carried around.

However, let's look at the case in question, sometimes we may want to prevent computing the length of the string multiple times even if that is a very optimized function (back in the day x86 had instructions like `scasb` but today this could be different but rest assured that the lib will try to use the most optimized instructions to do it).

This ends up calling `sinsp_utils::concatenate_paths`. The function does not really use the length of the string to decide how much data to copy, it recomputes it further down the stack but it does check if the length is `== 0` to verify if the string is empty. If we were to use `parinfo->m_len` an empty string will have length 1 and would skip the check, meaning that the previous use was not correct. Now, this is harmless since if you read the code the function will handle an empty string just fine but it's another reason why we shouldn't keep the original behavior in this case ;)

Reply:
Thanks for the in-depth explanation; i agree now!